qutip/core/data/convert.pyx

```diff
@@ -260,61 +262,159 @@ cdef class _to:
         for dispatcher in self.dispatchers:
             dispatcher.rebuild_lookup()
 
+    def register_alias(self, alias, layer_type):
+        """
+        Register a user frendly name for a data-layer type to be recognized by
+        the :method:`parse` method.
+
+        Parameters
+        ----------
+        alias : str
+            Name to be understood to represent the layer_type.
+
+        layer_type : type
+            Data-layer type, must have been registered with
+            :method:`add_conversions` first.
+        """
+        if layer_type not in self.dtypes:
+            raise ValueError(
+                "Type is not a data-layer type: " + repr(layer_type))
+        if type(alias) is not str:
+            raise TypeError("The alias must be a str")
+        self._str2type[alias] = layer_type
+
+    def parse(self, dtype):
+        """
+        Return a data-layer type object given its name or the type itself.
+
+        Parameters
+        ----------
+        dtype : type, str
+            Either the name of a data-layer type or a type itself.
+
+        Returns
+        -------
+        type
+            A data-layer type.
+
+        Raises
+        ------
+        TypeError
+            If ``dtype`` is neither a string nor a type.
+        ValueError
+            If ``dtype`` is a name, but no data-layer type of that name is
+            registered, or if ``dtype`` is a type, but not a known data-layer
+            type.
+        """
+        if type(dtype) is type:
+            if dtype not in self.dtypes:
+                raise ValueError(
+                    "Type is not a data-layer type: " + repr(dtype))
+            return dtype
+        elif type(dtype) is str:
+            try:
+                return self._str2type[dtype]
+            except KeyError:
+                raise ValueError(
+                    "Type name is not known to the data-layer: " + repr(dtype)
+                    ) from None
+
+        raise TypeError(
+            "Invalid dtype is neither a type nor a type name: " + repr(dtype))
+
     @cython.boundscheck(False)
     @cython.wraparound(False)
     def __getitem__(self, arg):
-        if isinstance(arg, type):
+        if type(arg) is not tuple:
             arg = (arg,)
-        if not isinstance(arg, tuple) or not arg or len(arg) > 2:
+        if not arg or len(arg) > 2:
             raise KeyError(arg)
-        to_t = arg[0]
-        if to_t not in self.dtypes:
-            raise TypeError("to_type is not known: " + str(to_t))
+        to_t = self.parse(arg[0])
```

Comment:
I'm not sure - I think it's quite neat to have the exact same errors raised regardless of the call method.  The `__getitem__` is more used as an overload, so the syntaxes `to(arg, type)` or `to[in_type, out_type]` are really just the same thing.

Particularly with the alias system, I think it is quite nice to have the `TypeError` and `ValueError` separation maintained; the former is _always_ going to be because of programming error, but the latter may appear for typo'd aliases, or if you've accidentally forgotten to import a plugin you wanted.  That seems kind of helpful to me as a user, whereas `KeyError` kind of seems not quite right - the fact that the dispatch system truly is handled by a key-lookup is kind of an implementation detail.  Maybe I'm not the best person to ask, though - I know too much about the internals, so what's "expected" to me might not be what a user thinks.

Reply:
I feel like `__getitem__` is kind of expected to mostly raise `KeyError` since that's what the interface looks like from the outside, but we can always fix it if anyone complains. :)